### PR TITLE
[JENKINS-71087] create dedicated 404 error page

### PR DESCRIPTION
**Pull request by mawinter69**
Currently Jenkins serves the error page of the underlying servlet container when a page doesn't exist or the permissions deny the discovery of the page.
The user experience will be better if a page is returned that will look like all other pages including navigation back to the root. Add a hint that it might be a permission problem.

Tested following scenarios:
- User is logged in and accesses a non existing pages -> 404 page is served
- User is logged in and accesses a page for which he has no discover rights -> 404 page is served
- User is logged in and acceses a page for which he has discover nut no read rights -> AccessDenied page is shown (this hints that the user should login, though he is logged in, but that is something for another PR)
- User is logged out and accesses a page where anonymous has discover rights but overall read -> login page is shown
- User is logged out and accesses a page where anonymous has no discover rights but overall read -> 404 page is shown
- User is logged out and accesses a non existing page, anonymous has overall read -> 404 page is shown

Development scenario:
run a `mvn hpi:run` for a plugin:
did the same tests as above, instead of the new 404 page a page with stapler trace is shown when 
 
Things that are not optimal currently:
- The 404 page is served in a way that a user appears as not logged even though he was logged in before. This probably also implies that the used theme is the one that is globally defined.

See [JENKINS-71087](https://issues.jenkins.io/browse/JENKINS-71087)

### Proposed changelog entries

- Show a Jenkins styled error page when a page is not found

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7865"></a>

### CODE REVIEW

File: core/src/main/resources/jenkins/model/Jenkins/404.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+notFound=The requested page wasn’t found. <br/>\
+  This means the page really doesn’t exist or a lack of permissions to discover the corresponding page.
```
Comment (NotMyFault) [inline core/src/main/resources/jenkins/model/Jenkins/404.properties:2]: I don't think this message is helpful for the end user. If you lack permission to access a page, Jenkins doesn't yield a 404 to the end user, but displays the permission they're lacking.

**Review by NotMyFault — COMMENTED** [review]
It would be beneficial to add a back to home button, similar to [jenkins.io](https://www.jenkins.io/badPage)

File: core/src/main/resources/jenkins/model/Jenkins/404.jelly [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+-->
+
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+  <l:layout title="Jenkins" type="one-column">
```
Comment (janfaracik) [inline core/src/main/resources/jenkins/model/Jenkins/404.jelly:27]:
```suggestion
  <l:layout title="404" type="one-column">
```

A more descriptive title would be good.

File: core/src/main/resources/jenkins/model/Jenkins/404.jelly [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    <l:header />
+    <l:main-panel>
+        <h1 style="text-align: center">
+          <img src="${imagesURL}/rage.svg" height="179" width="154"/> <span style="font-size:50px"><st:nbsp/>404 ${%Not found}!</span>
+        </h1>
+        <div id="error-description">
+          <h2 style="text-align: center">${%notFound}</h2>
+      </div>
```
Comment (janfaracik) [inline core/src/main/resources/jenkins/model/Jenkins/404.jelly:35]: Having the styling in it's own LESS file would be better than setting it in the 'style' attribute.

File: core/src/main/resources/jenkins/model/Jenkins/404.properties (hunk 1 above)
Comment (mawinter69) [reply core/src/main/resources/jenkins/model/Jenkins/404.properties:2]: There are 2 permissions that are relevant when you access jobs/folders. The Item.READ permissions and the Item.DISCOVER permission. If you have READ you can access the job. If you have only DISCOVER you either get an accessdenied with the hint that you should log on (even when you are logged on). If you're not logged in, then you're redirected to the login page.
If you have neither permission you get a 404. 
So the message is accurate in my eyes

**Comment by mawinter69** [comment]
> It would be beneficial to add a back to home button, similar to [jenkins.io](https://www.jenkins.io/badPage)

The page renders the usual header with logo (which is a link) and also the breadcrumb bar.

File: core/src/main/resources/jenkins/model/Jenkins/404.jelly (hunk 3 above)
Comment (mawinter69) [reply core/src/main/resources/jenkins/model/Jenkins/404.jelly:35]: I just copied the [oops](https://weekly.ci.jenkins.io/oops) page basically (Do you know that page?) So maybe there is also room to improve that page then.

**Comment by daniel-beck** [comment]
FTR https://github.com/jenkinsci/jenkins/pull/5527 was my attempt at doing this.

I don't remember what exactly blocked my PR at the time, but I think it was that logged-in users were not identified as such in the menu, since the 404 error pages happen outside the authentication filter.

Also, I had some test changes, perhaps those can address the test failures here?

**Comment by mawinter69** [comment]
Maybe we can revive your PR and just allow the images and adjuncts urls in the resorucedomainfilter so the page renders nice  there as well.

**Comment by daniel-beck** [comment]
https://github.com/daniel-beck/jenkins/commit/21c76e76847523f537c2fcade3630a296adbfd29 might be a solution to allow rendering the error page with knowledge of the current user (minimal testing so far), WDYT?

Ultimately the error page has no filters applied, and I don't know what kind of trouble that can cause. So I wonder whether we need to give up the idea of having an `l:layout` based error page, instead perhaps going with a more basic page.

**Comment by mawinter69** [comment]
That solves the issue with the user not being logged in. Not yet solved is the static files access via the resource domain. Here access is still broken as the images and scripts are not loaded. While that could be easily solved (allow the paths via the resourcedomainfilter) all links are not working as they all point to the resource domain url.
So maybe a more basic page is the way to go just showing the error message the image and a link back to the dashboard

**Comment by daniel-beck** [comment]
> Not yet solved is the static files access via the resource domain.

Would attempt to exclude the resource domain from the fancy error page using a similar mechanism. Perhaps have a "basic" 404 page in addition to the nice one, and if a new resource domain request attribute is set, render that instead.

**Comment by mawinter69** [comment]
created https://github.com/daniel-beck/jenkins/pull/10 in your branch that will handle the resource domain

**Comment by mawinter69** [comment]
Closing in favor of #8250

**Pull request closed without merging**